qutip/tests/test_gates.py

```diff
@@ -31,430 +31,329 @@
+        entangled: tensor-product ket -- the full entangled subspace
+        entangled_locations: list of int --
+            The locations that the qubits in the entangled subspace should be
+            in in the final tensor-product space.
+    """
+    n_entangled = len(entangled.dims[0])
+    n_separable = len(all_qubits) - n_entangled
+    separable = all_qubits.copy()
+    # Remove in reverse order so subsequent deletion locations don't change.
+    for location in sorted(entangled_locations, reverse=True):
+        del separable[location]
+    # Can't separate out entangled states to pass to tensor in the right places
+    # immediately, so tensor in at one end and then permute into place.
+    out = qutip.tensor(*separable, entangled)
+    permutation = list(range(n_separable))
+    current_locations = range(n_separable, n_separable + n_entangled)
+    # Sort to prevert later insertions changing previous locations.
+    insertions = sorted(zip(entangled_locations, current_locations),
+                        key=lambda x: x[0])
+    for out_location, current_location in insertions:
+        permutation.insert(out_location, current_location)
+    return out.permute(permutation)
+
+
+def _apply_permutation(permutation):
+    """
+    Permute the given permutation into the order denoted by its elements, i.e.
+        out[0] = permutation[permutation[0]]
+        out[1] = permutation[permutation[1]]
+        ...
 
-            ops = [basis(2, 1).dag(), basis(2, 0).dag(), identity(2)]
-            P = tensor(ops[p[0]], ops[p[1]], ops[p[2]])
-            assert_(P * U * P.dag() == identity(2))
+    This function is its own inverse.
+    """
+    out = [None] * len(permutation)
+    for value, location in enumerate(permutation):
+        out[location] = value
+    return out
+
+
+class TestExplicitForm:
+    def test_swap(self):
+        states = [qutip.rand_ket(2) for _ in [None]*2]
+        start = qutip.tensor(states)
+        swapped = qutip.tensor(states[::-1])
+        swap = gates.swap()
+        assert _infidelity(swapped, swap*start) < 1e-12
+        assert _infidelity(start, swap*swap*start) < 1e-12
+
+    @pytest.mark.parametrize('permutation', itertools.permutations([0, 1, 2]),
+                             ids=_permutation_id)
+    def test_toffoli(self, permutation):
+        test = gates.toffoli(N=3,
+                             controls=permutation[:2],
+                             target=permutation[2])
+        base = (qutip.tensor(1 - qutip.basis([2, 2], [1, 1]).proj(),
+                             qutip.qeye(2))
+                + qutip.tensor(qutip.basis([2, 2], [1, 1]).proj(),
+                               qutip.sigmax()))
+        # Iterate the permutation once, equivalent to finding its inverse,
+        # because for us, `permutation[n]` is where qubit `n` should be placed,
+        # whereas `Qobj.permuate` interprets that qubit `n` should be at the
+        # location `i` where `permutation[i] == n`.
```

Comment:
Yeah - that's somewhat deliberate.  I reuse the `_apply_permutation` function to generate tests for `expand_operator()` further down, since hopefully the code is very clear about what's happening.

The comment here is because the old test of the Toffoli gate had some very _very_ convoluted logic (that I don't even think was correct in the end...), and it took me a long time to work out what it was trying to do.